feat: Lazy-started transactions

brettwillis opened the pull request:
Eliminates at least one round-trip network request for every transaction by the following
- Transactions are started lazily upon the first read request (or not at all if no reads) instead of by the initial separate `beginTransaction` request. All subsequent reads (and commit/rollback) are queued upon the first read operation of each transaction attempt.
- Empty transactions (no reads or writes) become a network no-op (zero requests, no begin and no commit)
- ~Eliminate superfluous rollback request for some types of commit errors (where the transaction is already aborted by the server)~
- Added an internal `getResponse` API to the `DocumentReader`, `Query` and `AggregateQuery` classes to expose the transaction ID in the response.
- The internal `_stream()` APIs now emit a "transaction response" as the first element when appropriate. 

Other opportunistic improvements
- Avoid initialising useless write batch and backoff instances on read-only transactions

Fixes #2015

brettwillis committed:
Preliminary implementation of lazy transactions

brettwillis commented on dev/src/transaction.ts:
```diff
@@ -449,57 +472,46 @@ export class Transaction implements firestore.Transaction {
+      this._transactionIdPromise = undefined;
+      this._prevTransactionId = transactionId;
+    } catch (err) {
+      // A failed commit also rolls back the transaction so if we know that the
+      // server responded with an error then we clear the transaction ID so that
+      // we don't continue on to perform a superfluous rollback
+      this._transactionIdPromise = undefined;
+      this._prevTransactionId = transactionId;
+      throw err;
```
I feel that this statement is true just from my experimentation with the API, but it is not documented... is it correct? I need to determine if the error was an "error response" from the server and not some other comms error for example, what is the best way to determine this?

brettwillis committed:
Avoid creating new function contexts where possible

brettwillis committed:
Fixes an update tests

brettwillis committed:
Merge branch 'main' into transaction-lazy-begin

tom-andersen commented on dev/src/reference.ts:
```diff
@@ -2343,28 +2348,49 @@ export class Query<
-  _get(
-    transactionIdOrReadTime?: Uint8Array | Timestamp
+  async _get(
+    transactionOrReadTime?: Uint8Array | Timestamp | api.ITransactionOptions
```
Since the return type doesn't afford getting the new transactionId, I don't think we should allow passing in transaction options.

```suggestion
    transactionOrReadTime?: Uint8Array | Timestamp
```

brettwillis replied:
Done, btw.

brettwillis replied:
Done, btw.

brettwillis replied:
Done, btw.

tom-andersen reviewed (changes requested):
Thank you for the quite large contribution. I prefer if the rollback optimization based on error code was moved to a separate PR. I am not convinced that this change is correct. Verifying will hold up this PR. The problem is with reads having created locks, that should be released. If we wait for timeout of transaction, that can create problems for retry and other clients working on same documents.

I am giving you some preliminary feedback, but will return to do a more thorough review.

I sent you a PR to change the tests. For `readTime`, we do not need a transaction at all. This will eliminate the need to wait for first request to complete with transactionId, and this will also overcome potential transaction timeout.

brettwillis commented on dev/test/transaction.ts:
```diff
@@ -132,62 +132,69 @@ function rollback(
+    request.newTransaction = {
+      readOnly: options.newTransaction.readOnly.readTime
+        ? {
+            readTime:
```
Ok yes this makes sense, subsequent reads will not need to be queued. Will do.

brettwillis commented on dev/test/transaction.ts:
```diff
@@ -377,75 +423,72 @@ function runTransaction<T>(
+    [Status.FAILED_PRECONDITION]: {retry: false, rollback: false},
+    [Status.OUT_OF_RANGE]: {retry: false, rollback: true},
+    [Status.UNIMPLEMENTED]: {retry: false, rollback: true},
+    [Status.DATA_LOSS]: {retry: false, rollback: true},
```
Yes I think some thorough discussion is warranted on this. I can reverse it so that it can be implemented separately.

So this behaviour is not documented anywhere, but I discovered it via experimentation with the REST API, I'll show an example below.

But it kind of makes sense, here's an obvious example: if the commit fails saying "transaction has expired", then obviously there is no need to attempt rolling back an expired transaction.

However I've found is that the behaviour extends to several other error codes too.

Here's an example you can try yourself:

(1) Hold a lock on a document (non-existent document)

```
POST /v1/projects/test-project/databases/(default)/documents:batchGet HTTP/1.1
Host: firestore.googleapis.com
Authorization: Bearer •••••
Content-Type: application/json

{
    "newTransaction": {
        "readWrite": {}
    },
    "documents": [
        "projects/test-project/databases/(default)/documents/collection_id/non_existent_doc"
    ]
}
```

(2) Attempt to write that document outside the transaction. You will find that this request is blocked until the transaction from (1) completes for fails or times out.

```
POST /v1/projects/test-project/databases/(default)/documents:batchWrite HTTP/1.1
Host: firestore.googleapis.com
Authorization: Bearer •••••
Content-Type: application/json

{
    "writes": [
        {
            "update": {
                "name": "projects/test-project/databases/(default)/documents/collection_id/non_existent_doc",
                "fields": {}
            }
        }
    ]
}
```

(3) Commit the transaction from (1) with a precondition that will fail (i.e. precondition "exists" when the document doesn't exists causes the commit to fail).

```
POST /v1/projects/test-project/databases/(default)/documents:commit HTTP/1.1
Host: firestore.googleapis.com
Authorization: Bearer •••••
Content-Type: application/json

{
    "transaction": "<the transaction ID from (1)>"
    "writes": [
        {
            "update": {
                "name": "projects/test-project/databases/(default)/documents/collection_id/non_existent_doc",
                "fields": {}
            },
            "currentDocument": {
                "exists": true
            }
        }
    ]
}
```

The commit will fail with "NOT_FOUND" (no document to update). And in that moment, you will see that the request from (2) completes successfully meaning the failed commit released the lock on that document.

You'll need to complete this sequence within 20s or so otherwise (2) may time out anyway.

We'd definitely need input from the appropriate team on which error codes mean that the transaction has been aborted, and whether or not the behaviour is reliable. That is, it needs to be documented!

tom-andersen replied:
I received from clarification from Firestore. The rollback can be called in a non-blocking fashion. We do not need to wait for result of rollback before continuing.

I think making the rollback non-blocking will achieve the same latency as omitting the callback.

Feel free to implement this.

brettwillis replied:
Ok I get that. I'll remove those changes.

> The rollback can be called in a non-blocking fashion. We do not need to wait for result of rollback before continuing.

Hmm, but without knowing the specifics of the RPC or REST transport, what's to say that which of the rollback or the next transaction attempt requests arrive at the backend first? I.e. they could be on different streams or connections? Isn't that asking for a race condition?

Or is it that the next attempt would be blocked at the backend until the rollback is received anyway, so it doesn't matter which arrives first?

What will the backend do if/when it receives a `readWrite: { retryTransaction: "..." }` options to retry a transaction that hasn't been committed or rolled back yet? Does it automatically roll back the provided transaction?

Finally, regarding your point about errors from reads. If the initial read operation rejects, then as it is implemented currently we will obviously have no transaction ID to roll back. I was pondering whether or not it may be possible for:
1. The backend to lazily start the transaction and establish the lock, only for something in the SDK (e.g. validation) to reject, preventing us from receiving the transaction ID, and preventing us from rolling back the transaction that was established. I figured this was both highly unlikely and inescapable. 
2. The backend to send the "transaction ID response" in the response stream and then fail/error before sending all of "read responses" into the stream. In this case it is conceivable we could still have received a transaction ID even though the read operation finally rejects. With a bit of reworking we could capture the transaction ID and roll back the transaction even when the read rejects.

What are your thoughts on that?

tom-andersen replied:
The `retryTransaction` is more of a hint to backend, that can help process the transaction more efficiently. It won't create adverse behaviour if previous transaction is still active.

brettwillis replied:
Ok I've reverted the conditional rollback completely and made rollback complete asynchronously.

brettwillis replied:
Ok I've reverted the conditional rollback completely and made rollback complete asynchronously.

brettwillis committed:
Remove transaction options type from normal _get

brettwillis committed:
Revert rollback optimisation

brettwillis committed:
Do not start transaction when readTime specified

brettwillis committed:
Completely revert conditional rollback

brettwillis committed:
Rollback is completed asynchronously

tom-andersen commented on dev/src/transaction.ts:
```diff
@@ -622,17 +619,124 @@ export class Transaction implements firestore.Transaction {
+    transaction?: Uint8Array;
+    result: DocumentSnapshot<AppModelType, DbModelType>[];
+  }> {
+    const documentReader = new DocumentReader(this._firestore, documents);
```
I think the DocumentReader constructor should take `opts` as a third parameter. If you want, you can do this refactor.

brettwillis replied:
Hmm then wouldn't this also be the case for `readTime` and `fieldMask`? I do kind of think that all of these should actually be immutable options provided at construction rather than mutable fields that could be changed at any point before/during/after the request... Do you want me to change all of them to be provided in a single `opts` constructor parameter?

```
opts?: {
      transactionOrReadTime?: Uint8Array | api.ITransactionOptions | Timestamp;
      fieldMask?: FieldPath[];
    }
```

tom-andersen replied:
I completely agree that with making this immutable passed through constructor. Whether this should be an `opts` object or have `transactionOrReadTime` and `fieldMask` be parameters, I am inclined to go the parameter route, since that is the pattern used most other places in code. Regardless, what you implemented is a great improvement over previous code IMHO.

tom-andersen commented on dev/src/transaction.ts:
```diff
@@ -771,3 +875,24 @@ function isRetryableTransactionError(error: GoogleError): boolean {
+  await backoff.backoffAndWait();
+}
+
+function resolveResult<TResult>({result}: {result: TResult}): TResult {
```
This is a very small code snippet. I feel the indirection of calling this method obfuscates the code.

I think inlining this method with `v => v.result` is clearer and more concise.

brettwillis replied:
True

tom-andersen commented on dev/src/transaction.ts:
```diff
@@ -622,17 +619,124 @@ export class Transaction implements firestore.Transaction {
+      } else {
+        // This is the first read of the transaction so we create the appropriate
+        // options for lazily starting the transaction inside this first read op
+        let opts: api.ITransactionOptions;
+        if (this._writeBatch) {
+          opts = this._prevTransactionId
+            ? {readWrite: {retryTransaction: this._prevTransactionId}}
+            : {readWrite: {}};
+        } else {
+          opts = {readOnly: {}};
+        }
```
Not sure how much better this is, but I found this logic could maybe be written clearer. Your call. This is not a deal breaker.

```suggestion
        const opts: api.ITransactionOptions = {};
        if (this._writeBatch) {
          opts.readWrite = this._prevTransactionId
            ? {retryTransaction: this._prevTransactionId}
            : {};
        } else {
          opts.readOnly = {};
        }
```

brettwillis replied:
Sure

tom-andersen reviewed (changes requested):
Overall, I really like your implementation. There are a few things I would like to see changed, but then I see no reason why this shouldn't be approved and merged.

Note, there will be a delay before we can merge since this has to fit into our upcoming release schedule. We should have this out by April.

brettwillis commented on dev/src/transaction.ts:
```diff
@@ -557,15 +554,19 @@ export class Transaction implements firestore.Transaction {
         lastError = err;
 
-        if (!this._transactionId || !isRetryableTransactionError(err)) {
+        // We may or may not have a previous transaction ID
+        // For example, the initial read operation could have failed with a
+        // retry-able error meaning we did not obtain the ID of the lazy-started
+        // transaction so we will retry without the ID of the previous attempt
```
Ok sure will remove

brettwillis commented on dev/src/transaction.ts:
```diff
@@ -509,14 +491,28 @@ export class Transaction implements firestore.Transaction {
+    this._prevTransactionId = transactionId;
 
     try {
       await this._firestore.request('rollback', request, this._requestTag);
```
Yeah I have done so but this PR seems a bit broken at the moment GitHub is stuck processing my last commit for about a day now.

brettwillis commented on dev/src/document-reader.ts:
```diff
@@ -163,7 +197,8 @@ export class DocumentReader<AppModelType, DbModelType extends DocumentData> {
         // Transactional reads are retried via the transaction runner.
-        !this.transactionId &&
+        !this.transactionIdOrNewTransaction &&
+        !this.retrievedTransactionId &&
```
True.

tom-andersen replied:
Nice. I see you did one better in implementation.

brettwillis committed:
Cleanup

brettwillis commented:
There we go. With this last commit GitHub seems to have gotten itself unstuck. I've implemented all of those suggestions. You haven't seen the second-to-last commit `Rollback is completed asynchronously`, that was the one that was stuck processing.

brettwillis committed:
Fixes

Make resilient to wether transaction is included in same or different response
Test transaction ID buffer length

brettwillis commented:
@tom-andersen I ran this branch against a real Firestore backend and oops there were a few gotchas
- Protobuf includes zero/empty "transaction ID" buffers by default
- Sometimes the backend returns the "transaction" in the same response stream element as the "document response" and sometimes it doesn't which surprised me based on the docs. So I made it more resilient to which response comes and when.

Warrants another review sorry.

brettwillis committed:
Revert comment

brettwillis commented:
> Sometimes the backend returns the "transaction" in the same response stream element as the "document response" and sometimes it doesn't which surprised me based on the docs.

Ah sorry I think I was wrong about this... it was just the fact that protobuf always decodes an empty transaction ID buffer by default.

No harm in what I did to make it more resilient? Or do you want me to revert that?

By the way, will this branch be be run through a real-life test suite before releasing? In both REST and gRPC modes in case that affects the ordering/structuring of the responses?

tom-andersen commented:
> By the way, will this branch be be run through a real-life test suite before releasing? In both REST and gRPC modes in case that affects the ordering/structuring of the responses?

Yes. We have an integration tests that must pass before we merge. I hadn't run them yet, since there were still changes being done.

tom-andersen commented on dev/src/reference.ts:
```diff
@@ -3342,53 +3393,93 @@ export class AggregateQuery<
       stream.on('error', err => {
         reject(wrapError(err, stack));
       });
       stream.once('data', result => {
```
Integration tests are fixed by changing this:

```suggestion
      stream.on('data', result => {
```

tom-andersen commented:
> Ah sorry I think I was wrong about this... it was just the fact that protobuf always decodes an empty transaction ID buffer by default.
> 
> No harm in what I did to make it more resilient? Or do you want me to revert that?

I am fine with this.

brettwillis committed:
Fix aggregate query stream

Co-authored-by: Tom Andersen <tom-andersen@users.noreply.github.com>

brettwillis committed:
Apply suggestion for DocumentReader parameters

tom-andersen reviewed (approved):
LGTM

We are going to hold back on merging until it fits into our release cycle.

Fingers crossed that nothing causes a merge conflict.

brettwillis commented:
Thanks for all your time and effort reviewing this one @tom-andersen 🙏🏼

brettwillis committed:
Merge branch 'main' into transaction-lazy-begin

brettwillis committed:
Fix conformance tests

brettwillis commented:
@tom-andersen ok so those recent updates had quite sizeable conflict in `reference.ts`. I've resolved those, and using the `_getResonse()` pattern I introduced afforded quite a sizeable reduction in duplicated code related to "explain" functionality.

Now as a result I discovered something else. The Query response was silently accepting an undefined `readTime`. This was due to a shortfall in type checking. Due to one of my changes it surfaced in typings that `readTime` could be undefined, so I added a guard.

I turns out many tests were incorrectly simulating the "empty query" scenario by not emitting a read time in the response stream. So I made all "empty query" tests always emit a read time in the stream, which properly emulates how the backend should behave.

TLDR; I introduced a behaviour change wherein, if for whatever reason the backend did not emit a read time for query responses, it would previously succeed but the `QuerySnapshot.readTime` would be unexpectedly `undefined` in application code. Now it throws an exception in such case (which should not ever happen because the backend always emits a read time as far as I understand).

Thoughts?

brettwillis commented on dev/src/reference.ts:
```diff
@@ -1725,22 +1775,28 @@ class QueryUtil<
-                  changes.push(new DocumentChange('added', docs[i], -1, i));
-                }
-                return changes;
+          if (!output.readTime) {
+            reject(wrapError(Error('No read time'), stack));
+            return;
+          }
```
@tom-andersen  see here

tom-andersen replied:
Comment this out for now, with explanation that 'ExplainOptions' with 'analyze = false` will not have read time.

This is because no results are results are returned in this situation.

tom-andersen commented:
@brettwillis We are now within a window of time where I can merge this.

It is possible to return a query plan without any results. Consequently, there is no read time. This does not align with the types in the type system.

I would like @ehsannas who implemented this feature to take a look.

For now, since this is somewhat unrelated to your changes, I suggest we comment out the check, with an explanation that @ehsannas can follow up upon.

There are integration tests that I run as part of this PR that fail because the query plan does not return a read time.

> @tom-andersen ok so those recent updates had quite sizeable conflict in `reference.ts`. I've resolved those, and using the `_getResonse()` pattern I introduced afforded quite a sizeable reduction in duplicated code related to "explain" functionality.
> 
> Now as a result I discovered something else. The Query response was silently accepting an undefined `readTime`. This was due to a shortfall in type checking. Due to one of my changes it surfaced in typings that `readTime` could be undefined, so I added a guard.
> 
> I turns out many tests were incorrectly simulating the "empty query" scenario by not emitting a read time in the response stream. So I made all "empty query" tests always emit a read time in the stream, which properly emulates how the backend should behave.
> 
> TLDR; I introduced a behaviour change wherein, if for whatever reason the backend did not emit a read time for query responses, it would previously succeed but the `QuerySnapshot.readTime` would be unexpectedly `undefined` in application code. Now it throws an exception in such case (which should not ever happen because the backend always emits a read time as far as I understand).
> 
> Thoughts?

brettwillis committed:
Revert readTime null assertion behaviour

brettwillis commented:
Ok I reverted back to null assertion instead of throwing an error. This is back to the behaviour it has always been. And I added to-dos for @ehsannas to follow up in a separate PR.

tom-andersen commented on dev/src/reference.ts:
```diff
@@ -2982,99 +3048,62 @@ export class Query<
-            metrics = data.explainMetrics;
-
-            if (docs === null && metrics?.executionStats !== null) {
-              // This indicates that the query was executed, but no documents
```
This logic wasn't moved over. 

The effect is that returned snapshot will be null if there are no docs and `explainMetrics !== null` and `explainMetrics.executionStats === null`.

When `options.analyze !== true`, then only the `planSummary` is returned, but no execution actually happens. That means there will be no results, and no `executionStats`.

This is why the type `ExplainResults` has `readonly snapshot: T | null`

brettwillis replied:
Maybe I was mistaken, but this seemed like a just a long-winded way of achieving what I had instead which was just to initialise `docs` with an empty array to begin with.

I'm about to push a commit which changes the logic to only create a query snapshot if the stream received a `readTime` (and `docs` is still initialised to an empty array). So if the backend sends a readTime, when we get out a query snapshot with the received docs (or empty docs). If no read time received then no snapshot is created, and that should only happen with explain && analyze !== true (but is the prerogative of the backend rather than validating it in the lower levels of the SDK).

I then delegated further up the chain in the `_get()` function to verify that there is indeed a query snapshot present.

I also rationalised the `_get()`, `_getResponse()` and `_getResponseOrExplain()` methods:
- the internal `_get()` now returns response with metadata (and verifies that query snapshot exists). Previously this was just a trivial wrapper underneath the public `get()` of seemingly no use
- `_getReponse()` is the same but without the check for existence of a snapshot, and this is reused by `explain()`
- `_getResponseOrExplain()` removed
- Renamed `getResponse()` on `DocumentReader` to `_get()` to align with this convention

ehsannas replied:
SGTM

tom-andersen commented on dev/src/reference.ts:
```diff
@@ -1725,22 +1775,33 @@ class QueryUtil<
+          //   return;
+          // }
+
+          const result = query._createSnapshot(
```
`result` needs to be `null` when `options.analyze !== true`

The backend should only return `explainMetrics` in this case.

However, if `options.analyze === true`, then returning no docs is an empty result, and we should return a snapshot with no docs.

tom-andersen commented on dev/src/reference.ts:
```diff
@@ -1725,22 +1775,33 @@ class QueryUtil<
-                  changes.push(new DocumentChange('added', docs[i], -1, i));
-                }
-                return changes;
+          // TODO(ehsannas): Typings of QuerySnapshot suggest that readTime is always available
```
I talked to @ehsannas and read through some documentation.

```
// If the query returns no results, a response with `read_time` and no
// `document` will be sent, and this represents the time at which the query
// was run.
```
 
 Note, `readTime` represents the time at which query was run. So, when calling `explain(options)` with `options.analyze !== true`, indicating NO execution is to be done, ONLY a plan is to be returned, then `readTime` will be absent, but we also omit the snapshot from result.

tom-andersen commented on dev/src/reference.ts:
```diff
@@ -1458,6 +1458,33 @@ export class VectorQuerySnapshot<
+  DbModelType extends firestore.DocumentData = firestore.DocumentData,
+> {
+  transaction?: Uint8Array;
+  readTime?: Timestamp;
```
I don't think `readTime` ever exists outside the context of a snapshot. The snapshot should have a `readTime` property. So, we can remove from this object IMHO.

brettwillis replied:
This `QueryStreamResponse` represents the shape of any single "element" emitted from the stream. It is possible to the stream to emit only a `readTime` and no documents (empty result). So I think keeping `readTime` in this particular place is valid.

tom-andersen replied:
👍 I stand corrected!

tom-andersen commented on dev/src/reference.ts:
```diff
@@ -1458,6 +1458,33 @@ export class VectorQuerySnapshot<
+
+interface QueryResponse<TSnapshot> {
+  transaction?: Uint8Array;
+  readTime?: Timestamp;
```
Likewise, I think we can remove `readTime`, and only have it exist within the snapshot.

brettwillis replied:
Ok I agree, your comment from above does apply in this `QueryResponse` type. Will do.

tom-andersen commented on dev/src/reference.ts:
```diff
@@ -1458,6 +1458,33 @@ export class VectorQuerySnapshot<
+  transaction?: Uint8Array;
+  readTime?: Timestamp;
+  explainMetrics?: ExplainMetrics;
+  result: TSnapshot;
```
The `result` snapshot is optional. If running `explain(..)` without execution, it should be null. However, it is also possible to run `explain(..)` with execution results.

tom-andersen commented on dev/src/reference.ts:
```diff
@@ -2982,99 +3048,62 @@ export class Query<
+    }
+    return new ExplainResults(
+      explainMetrics,
+      result as QuerySnapshot<AppModelType, DbModelType>
```
`result` is `null` if `analyze !== true`

tom-andersen reviewed (changes requested):
I spoke to @ehsannas.

We should fix this up before merging and remove comment.

I have added comments to help guide change. Basically, if there is a `snapshot`, there will also be a `readTime`. However, we do not always want to produce a `snapshot` if `explain(..)` only returns a plan, without doing any execution.

Absence of documents is NOT the same as an empty result. It can also mean that NO execution was done as part of `explain`. The logic needs to respect this distinction and only return a snapshot for normal queries without explain, or explain query with `option.analyze === true`.

This does present a problem for the type system, since we are plucking values from a stream. So some coercion will be required.

I really appreciate your effort on this PR, and I genuinely feel bad that your work conflicted with two other features we were releasing.

brettwillis committed:
Fix query snapshot readTime logic

tom-andersen commented on dev/src/reference.ts:
```diff
@@ -4179,52 +4231,21 @@ export class AggregateQuery<
-      options = {};
+    const {result, explainMetrics} = await this._getResponse(
+      undefined,
+      options
```
We need to pass an object to invoke explain.

```suggestion
      options || {}
```

tom-andersen committed:
Update dev/src/reference.ts

I am simply going to apply this change...

tom-andersen reviewed (approved):
I made one fix to your logic in order to pass System Tests. LGTM.

tom-andersen commented on dev/src/document-reader.ts:
```diff
@@ -142,28 +178,31 @@ export class DocumentReader<AppModelType, DbModelType extends DocumentData> {
             'DocumentReader.fetchDocuments',
             requestTag,
             'Document missing: %s',
             response.missing!
```
```suggestion
            response.missing
```

brettwillis committed:
Remove un-needed null assertion

Co-authored-by: Tom Andersen <tom-andersen@users.noreply.github.com>

ehsannas commented on dev/src/transaction.ts:
```diff
@@ -622,17 +619,123 @@ export class Transaction implements firestore.Transaction {
+      // Simply queue this subsequent read operation after the first read
+      // operation has resolved and we don't expect a transaction ID in the
+      // response because we are not starting a new transaction
+      return this._transactionIdPromise.then(async opts => {
+        const r = await resultFn.call(this, param, opts);
+        return r.result;
+      });
```
nit: to make it more consistent with the rest of this function

```js
      return this._transactionIdPromise
                 .then(opts => resultFn.call(this, param, opts))
                 .then(r => r.result);
```

tom-andersen commented on dev/test/transaction.ts:
```diff
@@ -500,25 +558,21 @@ describe('failed transactions', () => {
-          rollback('foo1'),
+          query({newTransaction: {readWrite: {}}, error: serverError}),
+          // No rollback because the lazy-start operation failed
+          //rollback('foo1'),
```
There is no transaction id yet, since that is part of the result from first query. So we can't rollback, even if we wanted to.

ehsannas replied:
yup. I'm just saying we should remove the commented code.

tom-andersen replied:
Ahh. Good point! +1

MarkDuckworth reviewed (commented):
This is fantastic. There are many nice code improvements, but it's also a significant community contribution. Thank you for your efforts.

I have a few additional comments for your consideration. But otherwise I think it looks good.

The only potential adverse affect would be if a customer is implementing parallel reads in a transaction using `get()`, then this has a possibility to slow the transaction if the first read is slow/large-doc. But in many cases that could be mitigated with a call to `getAll()`.

ehsannas commented on dev/src/transaction.ts:
```diff
@@ -449,85 +447,87 @@ export class Transaction implements firestore.Transaction {
+
+    // Rollback request is completed asynchronously
+    // We don't need to wait for it to completed before continuing the next attempt
+    this._firestore
```
good point. Yes, I think it's safer to change this to the old behavior. In particular, this affects cases where the customer code is running inside a Cloud Function environment. The Cloud Function could terminate and the response for this rpc may arrive after the Cloud Function termination, which results in errors or warnings in the logs.

brettwillis replied:
Ok, will revert

ehsannas replied:
👍  That sounds fine to me.

ehsannas reviewed (approved):
Thanks for your contribution @brettwillis . LGTM -- only the remaining minor comments should be addressed.

brettwillis commented on dev/src/transaction.ts:
```diff
@@ -290,7 +288,7 @@ export class Transaction implements firestore.Transaction {
     data: firestore.WithFieldValue<AppModelType>
   ): Transaction {
-    if (this._readOnly) {
+    if (!this._writeBatch) {
```
@MarkDuckworth I have changed the logic a little, the `_readOnly` field no longer exists, with the following motivation
- `_writeBatch` is now initialised only when it is read-write to avoid initialising a write batch for readonly transactions (same for `_backoff`)
- then the `_readOnly` flag would be superfluous as the presence of `_writeBatch` is now the authoritative flag for whether it is readonly or read/write (so I removed `_readOnly`)
- and `_readOnlyReadTime` field is optional, just because a read time is not set doesn't mean it isn't readonly

MarkDuckworth replied:
Thanks for the detailed reply. Sounds good then.

brettwillis committed:
Apply suggested tweaks

tom-andersen committed:
Merge branch 'main' into transaction-lazy-begin
